packages/server/src/api/muster/muster.compliance.ts

```diff
@@ -45,29 +46,71 @@ export function toMusterIntermediateCompliance(rosters: Roster[]): IntermediateM
  * TBD later....
  * @param observations
  * @param musteringOpportunities
- * @param musterComplianceReport
+ * @param rosterForMusterCompliance
  */
-export function calculateMusterCompliance(
-  observations: RecordedObservations[],
-  musteringOpportunities: MusteringOpportunities,
-  musterComplianceReport: IntermediateMusterCompliance[],
-): MusterCompliance[] {
-  return [];
+export function calculateMusterCompliance(observations: RecordedObservations[], musteringOpportunities: MusteringOpportunities, rosterForMusterCompliance: RosterForMusterCompliance[]): MusterCompliance[] {
+  return rosterForMusterCompliance.map(roster => {
+    const edipi = roster.edipi;
+    const unitId = roster.unitId;
+    const unitMusteringOpportunities = musteringOpportunities[unitId];
+
+    if (!unitMusteringOpportunities || unitMusteringOpportunities.length === 0) {
+      return { totalMusters: 0, mustersReported: 0, musterPercent: 100, ...roster };
+    }
+
+    const totalMusters = unitMusteringOpportunities.length;
+
+    const edipiObservations = filterObservationsByEdipi(observations, edipi);
+    const mustersReported = calculateIndividualMusterCompliance(edipiObservations, unitMusteringOpportunities);
+    const musterPercent = (100 * mustersReported) / totalMusters;
+
+
+    return { totalMusters, mustersReported, musterPercent, ...roster };
+  });
+}
+
+function calculateIndividualMusterCompliance(observations: RecordedObservationTimestamp[], musteringOpportunities: MusterWindowWithStartAndEndDate[]): number {
+  let compliance = 0;
+  musteringOpportunities.forEach(musteringOpportunity => {
+    compliance += calculateCompliance(observations, musteringOpportunity.startMusterDate, musteringOpportunity.endMusterDate);
+  });
+  return compliance;
 }
 
+function filterObservationsByEdipi(observations: RecordedObservations[], edipi: string): RecordedObservationTimestamp[] {
+  return observations
+    .filter(observation => observation.edipi === edipi)
+    .map(observation => { return { timestamp: observation.timestamp }; });
+}
+
+function calculateCompliance(observationTimestamps: RecordedObservationTimestamp[], startMusterDate: Moment, endMusterDate: Moment): number {
```

Comment:
Between this method and the filtering above I get the sense we are doing a lot of stuff client-side (from DB perspective). Is it not possible to move some of this stuff directly into queries?

Reply:
Sure a different approach could have been taken but how do we know which one is better? What kind of criteria we would use?

Reply:
The general rule of thumb is that putting the responsibility on the DB will always be faster. For one the data is indexed, so you are not querying against values in a scanning/iterative fashion like here (at the very least things will be log(n) on the DB). The more items that are returned from the DB results in more memory being utilized on the server. Also in the case of a large number of rows being returned from the DB, your throughput will suffer as well (there is still a network connection from server to DB).

May not be a big deal since we are showing a limited number of days in the UI but if the data needs to be exported or larger queries need to be executed, it will not scale well. We can see what Austin says :)

Reply:
I think it is hard to tell which method would be more efficient and easier to maintain especially looking at a small subset of the code.

Reply:
The performance improvements are now captured here: https://app.asana.com/0/1188940305683068/1200767809950586/f